packages/lit-html/src/directives/async-replace.ts

```diff
@@ -7,72 +7,82 @@
+import {DisconnectableAwaiter} from './disconnectable-awaiter.js';
 
 type Mapper<T> = (v: T, index?: number) => unknown;
 
-class AsyncReplaceDirective extends AsyncDirective {
-  private _value?: AsyncIterable<unknown>;
-  private _reconnectResolver?: () => void;
-  private _reconnectPromise?: Promise<void>;
+export class AsyncReplaceDirective extends AsyncDirective {
+  private __iterable!: AsyncIterable<unknown>;
+  private __iterator!: AsyncIterator<unknown>;
+  protected index = 0;
+  private __nextAwaiter?: DisconnectableAwaiter<this, IteratorResult<unknown>>;
+  private __mapper?: Mapper<unknown>;
 
   // @ts-expect-error value not used, but we want a nice parameter for docs
   // eslint-disable-next-line @typescript-eslint/no-unused-vars
   render<T>(value: AsyncIterable<T>, _mapper?: Mapper<T>) {
     return noChange;
   }
 
-  update(_part: ChildPart, [value, mapper]: DirectiveParameters<this>) {
+  update(_part: ChildPart, [iterable, mapper]: DirectiveParameters<this>) {
     // If we've already set up this particular iterable, we don't need
     // to do anything.
-    if (value === this._value) {
+    if (iterable === this.__iterable) {
       return;
     }
-    this._value = value;
-    this.__iterate(mapper);
+    // If we're starting a new promise, stop tracking the last one, since its
+    // result is no longer relevant
+    this.__nextAwaiter?.disconnect();
+    this.__mapper = mapper;
+    this.__iterable = iterable;
+    this.index = 0;
+    // Note, we use the iterator protocol manually (rather than for/await/of),
+    // which lets us use the `DisconnectableAwaiter` helper to disconnect &
+    // reconnect awaiting the next value from the iterable
+    this.__iterator = iterable[Symbol.asyncIterator]();
+    this.__awaitNextValue();
     return noChange;
   }
 
-  // Separate function to avoid an iffe in update; update can't be async
-  // because its return value must be `noChange`
-  private async __iterate(mapper?: Mapper<unknown>) {
-    let i = 0;
-    const {_value: value} = this;
-    for await (let v of value!) {
-      // Check to make sure that value is the still the current value of
-      // the part, and if not bail because a new value owns this part
-      if (this._value !== value) {
-        break;
-      }
-
-      // If we were disconnected, pause until reconnected
-      if (this._reconnectPromise) {
-        await this._reconnectPromise;
-      }
-
-      // As a convenience, because functional-programming-style
-      // transforms of iterables and async iterables requires a library,
-      // we accept a mapper function. This is especially convenient for
-      // rendering a template for each item.
-      if (mapper !== undefined) {
-        v = mapper(v, i);
-      }
+  private __awaitNextValue() {
+    this.__nextAwaiter = new DisconnectableAwaiter(
+      this.__iterator!.next(),
+      this,
+      (directive, result) => directive.__commitResult(result)
+    );
```

Comment:
This still seems to be a lot more complex than the alternative. Every loop requires a new object, 2 promises, and 2/3 closures.

The `AsyncPause` and `WeakRefWrapper` (could just be `PseudoWeakRef` without the fallback to real `WeakRef`) requires just 2 objects and a closure (all of which can be reused for every iteration) and internal promise allocation.

Reply:
Ok, I think that makes sense. Did another update to go with that approach. @justinfagnani @sorvell also PTAL